File: 2641-cousins-in-binary-tree-ii/2641-cousins-in-binary-tree-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode replaceValueInTree(TreeNode root) {
        if (root == null) return root;

        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        List<Integer> levelSum = new ArrayList<>();

        while (!q.isEmpty()) {
            int currLvlSum = 0;
            int n = q.size();
            for (int i = 0; i < n; i++) {
                TreeNode curr = q.poll();
                currLvlSum += curr.val;
                if (curr.left != null) {
                    q.offer(curr.left);
                }
                if (curr.right != null) {
                    q.offer(curr.right);
                }
            }
            levelSum.add(currLvlSum);
        }

        q.offer(root);
        root.val = 0;
        int level = 1;

        while (!q.isEmpty()) {
            int n = q.size();
            for (int i = 0; i < n; i++) {
                TreeNode curr = q.poll();
                int siblingSum = (curr.left != null ? curr.left.val : 0) + (curr.right != null ? curr.right.val : 0);

                if (curr.left != null) {
                    curr.left.val = levelSum.get(level) - siblingSum;
                    q.offer(curr.left);
                }
                if (curr.right != null) {
                    curr.right.val = levelSum.get(level) - siblingSum;
                    q.offer(curr.right);
                }
            }
            level++;
        }

        return root;
    }
}

File: 2641-cousins-in-binary-tree-ii/2641-cousins-in-binary-tree-ii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque

class Solution:
    def replaceValueInTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return root
        
        q = deque([root])
        level_sum = []

        while q:
            curr_lvl_sum = 0
            n = len(q)
            for _ in range(n):
                curr = q.popleft()
                curr_lvl_sum += curr.val
                if curr.left:
                    q.append(curr.left)
                if curr.right:
                    q.append(curr.right)
            level_sum.append(curr_lvl_sum)

        q.append(root)
        root.val = 0
        level = 1

        while q:
            n = len(q)
            for _ in range(n):
                curr = q.popleft()
                sibling_sum = (curr.left.val if curr.left else 0) + (curr.right.val if curr.right else 0)

                if curr.left:
                    curr.left.val = level_sum[level] - sibling_sum
                    q.append(curr.left)
                if curr.right:
                    curr.right.val = level_sum[level] - sibling_sum
                    q.append(curr.right)
            level += 1

        return root